### PR TITLE
Actor critic

### PR DESCRIPTION
**Pull request by dnddnjs**
add actor-critic agent for cartpole

### CODE REVIEW

File: 1-grid-world/1-policy-iteration/environment.py [hunk 1]
```diff
@@ -0,0 +1,244 @@
+import tkinter as tk
+from tkinter import Button
+import time
+import numpy as np
+from PIL import ImageTk, Image
+
+PhotoImage = ImageTk.PhotoImage
+UNIT = 100  # 픽셀 수
```
Comment (keon) [inline 1-grid-world/1-policy-iteration/environment.py:8]: english

File: 2-cartpole/1-dqn/train.py [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+        _, action = torch.max(qvalue, 1)
+        return action.numpy()[0]
+
+
+def update_target_model(net, target_net):
+    target_net.load_state_dict(net.state_dict())
+
+
+
+if __name__=="__main__":
```
Comment (keon) [inline 2-cartpole/1-dqn/train.py]: we should wrap this function into something like `main()` and let `if __name__=="__main__"` call that function. 
variables and functions declared in this if __name__ == "__main__" is `global`

File: 2-cartpole/1-dqn/train.py [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+    print('state size:', num_inputs)
+    print('action size:', num_actions)
+
+    net = Model(num_inputs, num_actions)
+    def weights_init(m):
+        classname = m.__class__.__name__
+        if classname.find('Linear') != -1:
+            torch.nn.init.xavier_uniform(m.weight)
+
+    net.apply(weights_init)
```
Comment (keon) [inline 2-cartpole/1-dqn/train.py]: i think we can embed `weights_init` inside `Model` class and call it inside `__init__`.

File: 2-cartpole/1-dqn/train.py [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+import sys
+import gym
+import random
+import argparse
+import numpy as np
+
+import torch
+import torch.optim as optim
+import torch.nn.functional as F
+from model import Model
```
Comment (keon) [inline 2-cartpole/1-dqn/train.py]: Can we have a better name for `Model`?

File: 2-cartpole/2-actor-critic/main.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+parser.add_argument('--render', default=False, action="store_true")
+parser.add_argument('--gamma', default=0.99, help='')
+parser.add_argument('--log_interval', default=10, help='')
+parser.add_argument('--logdir', type=str, default='./logs',
+                    help='tensorboardx logs directory')
+args = parser.parse_args()
+
+
+def train_model(sample, policy, value):
+    state = sample[0]
```
Comment (keon) [inline 2-cartpole/2-actor-critic/main.py:26]: we can replace this with one-liner
```
state, action, reward, next_state, mask = sample
```

File: 2-cartpole/2-actor-critic/main.py [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+    action = sample[1]
+    reward = sample[2]
+    next_state = sample[3]
+    mask = sample[4]
+
+    _, next_value = net(next_state)
+    pred = reward + mask * args.gamma * next_value[0]
```
Comment (keon) [inline 2-cartpole/2-actor-critic/main.py:33]: what is the purpose of `mask`?

File: 2-cartpole/2-actor-critic/test.py [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+args = parser.parse_args()
+
+
+def get_action(policy):
+    policy = policy.data.numpy()[0]
+    action = np.random.choice(num_actions, 1, p=policy)[0]
+    return action
+
+
+if __name__=="__main__":
```
Comment (keon) [inline 2-cartpole/2-actor-critic/test.py]: same applies for every files.
we should use `main()` and call it here.

File: 2-cartpole/2-actor-critic/main.py [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+            policy, value = net(state)
+            action = get_action(policy)
+
+            next_state, reward, done, _ = env.step(action)
+            next_state = torch.Tensor(next_state)
+            next_state = next_state.unsqueeze(0)
+            
+            mask = 0 if done else 1
+            reward = reward if not done or score == 499 else -1
+            sample = [state, action, reward, next_state, mask]
```
Comment (keon) [inline 2-cartpole/2-actor-critic/main.py:94]: what would be a better name for `sample`?

File: 2-cartpole/2-actor-critic/main.py [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+
+
+def train_model(sample, policy, value):
+    state = sample[0]
+    action = sample[1]
+    reward = sample[2]
+    next_state = sample[3]
+    mask = sample[4]
+
+    _, next_value = net(next_state)
```
Comment (keon) [inline 2-cartpole/2-actor-critic/main.py:32]: this is one example of the problem of declaring things inside `if __name__ == '__main__'`.
`net` was declared below as a global variable and we are able to use it here, which may confuse reader.
net should be one of the arguments of this function.

File: 2-cartpole/2-actor-critic/train.py [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+parser.add_argument('--gamma', default=0.99, help='')
+parser.add_argument('--goal_score', default=400, help='')
+parser.add_argument('--log_interval', default=10, help='')
+parser.add_argument('--logdir', type=str, default='./logs',
+                    help='tensorboardx logs directory')
+args = parser.parse_args()
+
+
+def train_model(sample, policy, value):
+    state = sample[0]
```
Comment (keon) [inline 2-cartpole/2-actor-critic/train.py]: same one-liner change can apply here.

File: 2-cartpole/2-actor-critic/train.py [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+    action = sample[1]
+    reward = sample[2]
+    next_state = sample[3]
+    mask = sample[4]
+
+    _, next_value = net(next_state)
```
Comment (keon) [inline 2-cartpole/2-actor-critic/train.py:30]: same problem as above

File: 2-cartpole/3-multi-step/train.py [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+    for e in range(3000):
+        done = False
+        score = 0
+        steps = 0
+
+        state = env.reset()
+        state = torch.Tensor(state)
+        state = state.unsqueeze(0)
+
+        states, next_states, actions, rewards, masks = [], [], [], [], []
```
Comment (keon) [inline 2-cartpole/3-multi-step/train.py]: there seems to be multiple patterns used for storing memory.
There's Memory class for dqn, and here we store state, next_state, actions, and etc into separate lists.
We should consolidate the style.

File: 3-atari/1-dqn/train.py [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+    env.seed(500)
+    torch.manual_seed(500)
+
+    img_shape = env.observation_space.shape
+    num_actions = 3
+    print('image size:', img_shape)
+    print('action size:', num_actions)
+
+    net = Model(num_actions)
+    def weights_init(m):
```
Comment (keon) [inline 3-atari/1-dqn/train.py]: this should be embedded into the model class as I commented for cartpole dqn

File: 3-atari/1-dqn/train.py [hunk 14]
```diff
@@ -0,0 +1,173 @@
...
+import sys
+import gym
+import random
+import argparse
+import numpy as np
+
+import torch
+import torch.optim as optim
+import torch.nn.functional as F
+from utils import *
```
Comment (keon) [inline 3-atari/1-dqn/train.py]: plz refrain from using import *.
it makes difficult to understand what comes from what.

File: 3-atari/1-dqn/train.py [hunk 15]
```diff
@@ -0,0 +1,173 @@
...
+parser.add_argument('--update_target', default=10000, help='')
+parser.add_argument('--log_interval', default=1, help='')
+parser.add_argument('--goal_score', default=300, help='')
+parser.add_argument('--logdir', type=str, default='./logs',
+                    help='tensorboardx logs directory')
+args = parser.parse_args()
+
+
+def train_model(batch):
+    history = torch.stack(batch.history).cuda()
```
Comment (keon) [inline 3-atari/1-dqn/train.py]: instead of `cuda`, we should use `to` function.

good pattern for this is:
```python
DEVICE = torch.device("cuda" if torch.cuda.is_available() else "cpu") # declare this at the top
t = t.to(DEVICE) # t is the torch tensor
```
this will automatically send t to "cuda" or "cpu" depending on the user's setting

File: 3-atari/1-dqn/train.py [hunk 16]
```diff
@@ -0,0 +1,173 @@
...
+
+        while not done:
+            if args.render:
+                env.render()
+
+            steps += 1
+            qvalue = net(history.unsqueeze(0))
+            action = get_action(epsilon, qvalue, num_actions)
+            
+            if action == 0:
```
Comment (keon) [inline 3-atari/1-dqn/train.py]: `real_action = action + 1` ?

Comment (keon) [reply 3-atari/1-dqn/train.py]: or better, add 1 inside `get_action`?

**Review by keon — CHANGES_REQUESTED** [review]
I added some comments. Comments are mostly about styles.
I couldn't review the algorithm in detail, so I'll comeback tomorrow to comment on the logic in detail.

File: 2-cartpole/1-dqn/train.py (hunk 4 above)
Comment (dnddnjs) [reply 2-cartpole/1-dqn/train.py]: Then maybe qnet is better name for the model

File: 2-cartpole/2-actor-critic/main.py (hunk 8 above)
Comment (dnddnjs) [reply 2-cartpole/2-actor-critic/main.py:94]: Transition or transition_tuple would be replacement for the sample. Or I can feed the sample directly to the memory.push() method

File: 2-cartpole/2-actor-critic/main.py (hunk 6 above)
Comment (dnddnjs) [reply 2-cartpole/2-actor-critic/main.py:33]: if the sample is from end of episode, then next_value has to be 0. mask is 0 if the sample is from end of episode

**Review by keon — APPROVED** [review]
